refactor(robot-server, api): Remove `definition` from `loadModule` result

mjhuff opened the pull request:
Closes [EXEC-212](https://opentrons.atlassian.net/browse/EXEC-212) and works toward [EXEC-756](https://opentrons.atlassian.net/browse/EXEC-756) and [EXEC-77](https://opentrons.atlassian.net/browse/EXEC-77)

<!--
Thanks for taking the time to open a Pull Request (PR)! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

GitHub provides robust markdown to format your PR. Links, diagrams, pictures, and videos along with text formatting make it possible to create a rich and informative PR. For more information on GitHub markdown, see:

https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview

The `loadModule` command returns the module `definition` as a part of the command result. This isn't great, because module definitions are not customer facing, and they create versioning risks. We're removing `definition` from `loadModule` command now because locating features will touch module definitions, and not removing definitions means we will have to handle a quite complex database migration as an alternative.

0d6d69c5c1e6158b83c928a5efa820337813257c is the beefy commit. There are really two simultaneously refactors occurring here. First, we remove `definition` from `load_module`, and handle all the lint errors that arise from that. Second, some internals within `ModuleState` were dependent upon that module definition supplied from the command result, so we handle this by creating a new `StateUpdate` mechanism for module loading. 

501b0010b746eb4e723ce8e286c3726607ceb519 is the database migration. See the comment in `v11_to_v12.py` for why this is necessary.

As a side note, the `LoadModuleResult` in `shared-data` already doesn't include the `definition`, so the FE shouldn't depend on it without any further changes.
<!--
Describe your PR at a high level. State acceptance criteria and how this PR fits into other work. Link issues, PRs, and other relevant resources.
-->

## Test Plan and Hands on Testing
- Should be covered by CI here.
<!--
Describe your testing of the PR. Emphasize testing not reflected in the code. Attach protocols, logs, screenshots and any other assets that support your testing.
-->

## Changelog
- Removed `definition` from the `result` of the `loadModule` command.
<!--
List changes introduced by this PR considering future developers and the end user. Give careful thought and clear documentation to breaking changes.
-->

## Review requests
- Nothing in particular. This is my first database migration and first time dealing with `StateUpdate`, so a code check would be appreciated.
<!--
- What do you need from reviewers to feel confident this PR is ready to merge?
- Ask questions.
-->

## Risk assessment
low with tests. this does do a db migration. Unless the FE is doing some type assertions for `definition`, there shouldn't be any FE risk.
<!--
- Indicate the level of attention this PR needs.
- Provide context to guide reviewers.
- Discuss trade-offs, coupling, and side effects.
- Look for the possibility, even if you think it's small, that your change may affect some other part of the system.
  - For instance, changing return tip behavior may also change the behavior of labware calibration.
- How do your unit tests and on hands on testing mitigate this PR's risks and the risk of future regressions?
- Especially in high risk PRs, explain how you know your testing is enough.
-->


[EXEC-212]: https://opentrons.atlassian.net/browse/EXEC-212?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[EXEC-756]: https://opentrons.atlassian.net/browse/EXEC-756?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[EXEC-77]: https://opentrons.atlassian.net/browse/EXEC-77?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

mjhuff committed:
refactor(api): remove module def from loadModule results

mjhuff committed:
feat(robot-server): migrate db from 11 to 12

SyntaxColoring commented on api/src/opentrons/protocol_engine/commands/load_module.py:
```diff
@@ -82,16 +81,6 @@ class LoadModuleResult(BaseModel):
         description="An ID to reference this module in subsequent commands."
     )
 
-    # TODO(mm, 2023-04-13): Remove this field. Jira RSS-221.
-    definition: ModuleDefinition = Field(
-        json_schema_extra={"deprecated": True},
-        description=(
-            "The definition of the connected module."
-            " This field is an implementation detail. We might change or remove it without warning."
-            " Do not access it or rely on it being present."
-        ),
-    )
-
```
Might be worth leaving a brief note that `definition` used to exist here. Just on the off chance that someone tries to add a new field with the same name (which would be dangerous) or something like that.

SyntaxColoring commented on api/src/opentrons/protocol_engine/state/update_types.py:
```diff
@@ -409,6 +411,17 @@ class AddressableAreaUsedUpdate:
+    definition: ModuleDefinition
+    serial_number: Optional[str]
+    slot_name: Optional[DeckSlotName]
+    requested_model: Optional[ModuleModel]
```
Nit:

Can we make `slot_name` and `requested_model` non-`Optional`, following their prior types from when `ModuleStore` looked directly at `LoadModuleParams`?

The internals of `ModuleStore` do handle these being `None`, but that's for the benefit of "stateless commands", i.e. the `/commands` endpoints, which are used to control modules outside of a run. Within a run, a command should always have non`-None` values for these fields.

(`serial_number` should stay `Optional` because that's how Magnetic Blocks work.)

mjhuff replied:
Ahhh yep, good call out

github-actions[bot] committed:
fix(analyses-snapshot-testing): heal update_load-module-cmd snapshots (#18640)

This PR was requested on the PR
https://github.com/Opentrons/opentrons/pull/18639

Co-authored-by: mjhuff <64858653+mjhuff@users.noreply.github.com>

SyntaxColoring commented on api/tests/opentrons/protocol_engine/state/test_geometry_view.py:
```diff
@@ -469,7 +478,6 @@ def load_module_action(
             status=CommandStatus.SUCCEEDED,
             result=LoadModuleResult(
                 moduleId=module_id,
-                definition=module_def,
```
Nit: I think we can replace this whole `LoadModule` object with `_dummy_command()` now..

SyntaxColoring commented on api/tests/opentrons/protocol_engine/commands/test_load_module.py:
```diff
@@ -187,7 +197,7 @@ async def test_load_module_raises_if_location_occupied(
     equipment: EquipmentHandler,
     state_view: StateView,
 ) -> None:
-    """A loadModule command should have an execution implementation."""
+    """A loadModule command should raise an error if the location is occupied."""
```
TY for docstring updates.

SyntaxColoring commented on api/tests/opentrons/protocol_engine/commands/test_load_module.py:
```diff
@@ -172,13 +173,22 @@ async def test_load_module_implementation(
+        serial_number="mod-serial",
+        slot_name=load_slot_name,
+    )
+
```
In tests, I have been writing state updates with the syntax:

```python
state_update = StateUpdate(
    loaded_module=...
)
```

Instead of:

```python
state_update = StateUpdate()
state_update.set_load_module(...)
```

The benefit to doing this is that it's a check against the production code accidentally trying something like this:

```python
state_update = StateUpdate()
state_update.set_load_module(...)  # Load one module.
state_update.set_load_module(...)  # Try to load a second module. Probably wrong--overwrites the first one.
```

This is a really nitty stylistic thing and I'm not even sure if I actually prefer it. Just bringing it up because you asked in your review requests.

mjhuff replied:
Ah thanks, I do see the benefit here.

SyntaxColoring commented on robot-server/robot_server/persistence/_migrations/v11_to_v12.py:
Yep, this looks perfect.

SyntaxColoring reviewed (approved):
Yaaay, thank you! LGTM once CI passes.

mjhuff committed:
tavern fixes

mjhuff committed:
feedback
